WIP - Supermarket Backup and Restore Update (#1861)
chef/chef-web-docs: chef_master/source/supermarket_backup_restore.rst

=====================================================
`[edit on GitHub] <https://github.com/chef/chef-web-docs/blob/master/chef_master/source/supermarket_backup_restore.rst>`__

Periodic backups of Supermarket data are an essential part of managing and maintaining a healthy configuration, and to help ensure that important data can be restored if required. In a typical installation of Supermarket, both the cookbook store and the database need to be backed up on a regular basis.

Backup
=====================================================

Cookbook Backup
-----------------------------------------------------

> If you are not using AWS S3 storage for cookbooks, you will also need to make a backup of the local cookbook storage location. 

Reviewer: I'm thinking that the cookbook backup section should come after the database backup section since it would only apply if they are not using S3.  

Also I think the wording here should be:

If Supermarket is not configured to use AWS S3 storage for cookbooks, then the local cookbook storage location on the Supermarket server will also need to be backed up.

> If you are not using AWS S3 storage for cookbooks, you will also need to make a backup of the local cookbook storage location. 

Reviewer 2: Maybe omit "also"?

```suggestion
If you are not using AWS S3 storage for cookbooks, you will need to make a backup of the local cookbook storage location. 
```

The default location is: ``/var/opt/supermarket/data/cookbook_versions``.

For example, a cookbook backup command:

.. code-block:: bash

     cd /var/opt/supermarket/data/ 

>      tar cvzf supermarket_cookbook_versions.tar.gz cookbook_versions

Reviewer 2: I recommend not putting the resulting backup tar.gz file in the `/var/opt/supermarket/data/ ` directory. Maybe this way, which will place it in the user's home directory:

```suggestion
     tar cvzf ~/supermarket_cookbook_versions.tar.gz cookbook_versions
```

Database Backup
-----------------------------------------------------

A database export can be made in several formats. 

For example, a database export in a .dump format can be made with the following syntax:

.. code-block:: bash

>      $ /opt/supermarket/embedded/bin/pg_dump -h localhost -U supermarket -d supermarket_production -p 15432 -F c -b -v -f smkt-date-blah-backup.dump

Reviewer: should change `supermarket_production` to just `supermarket` as that should be the default db name

>      $ /opt/supermarket/embedded/bin/pg_dump -h localhost -U supermarket -d supermarket_production -p 15432 -F c -b -v -f smkt-date-blah-backup.dump

Reviewer 2: Here, too, it might be good to ensure the backup is going to the user's home directory.

I've removed my silly "blah" and expanded the options to maybe help the command seem less arcane.

```suggestion
     $ /opt/supermarket/embedded/bin/pg_dump --host localhost --username supermarket --dbname supermarket --port 15432 --format c --blobs --verbose --file ~/supermarket_database_backup.dump
```

where, in a typical installation:

>   * ``/opt/supermarket/embedded/bin/pg_dump`` is

Author: Need to finish this thought!

Reviewer 2: ```suggestion
  * ``/opt/supermarket/embedded/bin/pg_dump`` is the path to the database export utility included in the Supermarket installation.
```

  * ``localhost`` may alternatively be 127.0.0.1 
  * ``15432`` is PostgreSQL port number, which may need to be modified.

>   * ``-F c`` makes a custom .dump format

Reviewer 2: To match the longer command switches in the example:

```suggestion
  * ``--format c`` sets the output to PostgreSQL's "custom" binary file format
```

For documentation about the pg_dump utility, see: https://www.postgresql.org/docs/9.3/app-pgdump.html

> Be sure to update the various local values in the syntax as necessary to match your infrastructure. To find local variables, look at ``/var/opt/supermarket/etc/database.yml``. For example, an output of ``database.yml`` may look like:

Reviewer: change the word "syntax" to "`pg_dump` command"

so it should read

Be sure to update the various values in the `pg_dump` command as necessary to match your infrastructure...

> Be sure to update the various local values in the syntax as necessary to match your infrastructure. To find local variables, look at ``/var/opt/supermarket/etc/database.yml``. For example, an output of ``database.yml`` may look like:

Reviewer 2: ```suggestion
Be sure to update the various local values in the syntax as necessary to match your infrastructure. To find local variables, look at ``/etc/supermarket/supermarket-running.json``. For example, within this file, the ``"database"`` key may look like:
```

Another option to locate the values they're really operating with is to look in `/etc/supermarket/supermarket-running.json`. This file lives next to `supermarket.rb` and `supermarket.json` where their configuration is _set_. `supermarket-running.json` are the final values the system is operating with after running `sudo supermarket-ctl reconfigure`.

There's a `"database"` key in there:

```json
{ "supermarket": {
  ...
  "database": {
    "user": "supermarket",
    "name": "supermarket",
    "host": "127.0.0.1",
    "port": "15432",
    "pool": "30",
    "extensions": {
      "plpgsql": true,
      "pg_trgm": true
    },
    "password": "sup3rs3cr3t"
  },
  ...
}
```

Author: @robbkidd Would you suggest this option in .json over the .yml one? Otherwise, I can place both of these as options.

.. code-block:: yaml

     production:
         adapter: postgresql
         database: supermarket
         username: supermarket
         password:
         host: 127.0.0.1
         port: 15432
         pool: 25

Restore
=====================================================

Cookbook Restore
-----------------------------------------------------

When restoring cookbooks, **make sure the cookbook directory is writable by the supermarket user.**

For example, to restore your cookbook files, run:

.. code-block:: bash

     cd /var/opt/supermarket/data/
      tar xvfz /path/to/supermarket_cookbook_versions.tar.gz

Database Restore
-----------------------------------------------------

.. note::

    The restore does not support transferring backups across different versions of Supermarket. Backups taken must be restored to the same version of Supermarket that was in use when they were created.

For example, to restore a backup in a .dump format, run: 

.. code-block:: bash

>      $ pg_restore -h localhost -p 15432 --clean --no-acl --no-owner -d supermarket_production -v smkt-date-blah-backup.dump

Reviewer: needs same change for `supermarket_production` to just `supermarket`

>      $ pg_restore -h localhost -p 15432 --clean --no-acl --no-owner -d supermarket_production -v smkt-date-blah-backup.dump

Reviewer 2: Long-form switches to ~reduce arcanitiousness~ ... ~reduce occultiness~ ... make it clearerer.

```suggestion
     $ pg_restore --host localhost --port 15432 --clean --no-acl --no-owner --dbname supermarket_production --verbose supermarket_database_backup.dump
```

Periodic backups of Supermarket data are an essential part of managing and maintaining a healthy configuration, and to help ensure that important data can be restored if required. In a typical installation of Supermarket, both the cookbook store and the database need to be backed up on a regular basis.

Backup
=====================================================

Cookbook Backup
-----------------------------------------------------

If Supermarket is not configured to use AWS S3 storage for cookbooks, then the local cookbook storage location on the Supermarket server will also need to be backed up. 

The default location is: ``/var/opt/supermarket/data/cookbook_versions``.

For example, a cookbook backup command:

.. code-block:: bash

     cd /var/opt/supermarket/data/ 
     tar cvzf ~/supermarket_cookbook_versions.tar.gz cookbook_versions

Database Backup
-----------------------------------------------------

A database export can be made in several formats. 

For example, a database export in a .dump format can be made with the following syntax:

.. code-block:: bash

>      $ $ /opt/supermarket/embedded/bin/pg_dump --host localhost --username supermarket --dbname supermarket --port 15432 --format c --blobs --verbose --file ~/supermarket_database_backup.dump

Reviewer 2: Looks like an extra `$` prompt snuck in here.

=====================================================
`[edit on GitHub] <https://github.com/chef/chef-web-docs/blob/master/chef_master/source/supermarket_backup_restore.rst>`__

Periodic backups of Supermarket data are an essential part of managing and maintaining a healthy configuration, and to help ensure that important data can be restored if required. In a typical installation of Supermarket, both the cookbook store and the database need to be backed up on a regular basis.

Backup
=====================================================

Cookbook Backup
-----------------------------------------------------

> If Supermarket is not configured to use AWS S3 storage for cookbooks, then the local cookbook storage location on the Supermarket server will also need to be backed up. 

Author: @robbkidd How is this?

Reviewer 2: I think this is an English thing that could be tripping me up (and that I could be wrong about) on the "also." At the point the reader has arrived at this page, they haven't been told to back up anything yet. But the "also" makes it seem like we're adding an item to a list that only just started. So I think this can read "... will need to be backed up."

Author: Fixed!

There's a "database" key in ``supermarket-running.json``:

.. code-block:: javascript

      { "supermarket": {
           ...
           "database": {
                "user": "supermarket",
                "name": "supermarket",
                "host": "127.0.0.1",
                "port": "15432",
                "pool": "30",
                "extensions": {
                     "plpgsql": true,
                     "pg_trgm": true
                },
                "password": "sup3rs3cr3t"
           },
           ...
      }

Restore
=====================================================

Cookbook Restore
-----------------------------------------------------

When restoring cookbooks, **make sure the cookbook directory is writable by the supermarket user.**

For example, to restore your cookbook files, run:

.. code-block:: bash

      $ cd /var/opt/supermarket/data/ 

>       $ tar cvzf /supermarket_cookbook_versions.tar.gz

Reviewer 2: D'oh! I missed this in the first review. 🤦‍♂️ Need to change the option passed to `tar`: `c` when **C**reating a tar file, `x` when e**X**panding a tar file.

```suggestion
      $ tar xvzf /supermarket_cookbook_versions.tar.gz
```
